designs/sam_publish_app_cmd.rst

```diff
@@ -0,0 +1,230 @@
+.. contents:: **Table of Contents**
+   :depth: 2
+   :local:
+
+``sam publish app`` command
+====================================
+
+This is the design for a command to publish an application to `AWS Serverless Application Repository (SAR)`_ with a SAM
+template. It can be used to create a new application and its first version, update exisitng application's metadata, create
+a new version of the application, and manage application permissions.
+
+.. _AWS Serverless Application Repository (SAR): https://aws.amazon.com/serverless/serverlessrepo/
+
+
+What is the problem?
+--------------------
+To publish an app to AWS Serverless Application Repository, customers need to go through the following steps: first upload
+the application code and SAM template to an Amazon S3 bucket, correctly set S3 bucket policy that grants the service read
+permissions for artifacts uploaded to S3, then open the AWS Serverless Application Repository console and provide information
+in a bunch of input boxes. If they use the AWS CLI, they need to pass all the information as parameters, and it's easy to make
+a mistake while typing in the command line.
+
+
+What will be changed?
+---------------------
+In this proposal, we will be providing a new command, ``sam publish app``, which takes a SAM template as input and publishes
+an application to AWS Serverless Application Repository using applicaiton metadata specified in the template. Customers just
+need to provide application metadata information in the template, then ``sam package`` will handle uploading local files to S3,
+and ``sam publish app`` will create the app in Serverless Application Repository. We will also provide sharing options to set
+application permission policies. This command will greatly simplify the exsiting publishing experience.
+
+
+Success criteria for the change
+-------------------------------
+#. Support all the following use cases:
+
+   * Create new application and its first version in SAR using ``sam publish app``
+   * Create new version of existing SAR application using ``sam publish app``
+   * Update application metadata of existing SAR application using ``sam publish app``
+   * Share the app publicly using the ``--make-public`` option
+   * Make the app private using the ``--make-private`` option
+   * Share the app privately with other AWS accounts using the ``--account-ids`` option
+
+
+#. ``sam package`` command can upload local readme/license files to S3.
+
+
+Out-of-Scope
+------------
+#. Manage application permission separately without publishing/updating the app.
+
+#. Specify granular permission types as defined in `application permission`_ when sharing the application.
+
+#. Recursively publish nested apps in the template (SAR CreateApplication API doesn't support yet).
```

Comment:
How will customers do this? We should clearly document how customers can accomplish this. 

If I try to publish a nested app, will the command fail in any way?

Reply:
Customers can use aws cli to do this: https://docs.aws.amazon.com/serverlessrepo/latest/devguide/access-control-resource-based.html#application-permissions. We want to simplify the sam command interface since usually people want to give all permissions when they share app.

The command will fail if the referred nested app hasn't been created yet. Customers have to provide application id to nest the app in template, and publish doesn't touch those apps.